mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -124,14 +124,34 @@ private[feature] trait StringIndexerBase extends Params with HasHandleInvalid wi
 
     val outputFields = inputColNames.zip(outputColNames).flatMap {
       case (inputColName, outputColName) =>
-        schema.fieldNames.contains(inputColName) match {
-          case true => Some(validateAndTransformField(schema, inputColName, outputColName))
-          case false if skipNonExistsCol => None
+        extractInputDataType(schema, inputColName) match {
+          case Some(dtype) => Some(
+            validateAndTransformField(schema, inputColName, dtype, outputColName)
+          )
+          case None if skipNonExistsCol => None
           case _ => throw new SparkException(s"Input column $inputColName does not exist.")
         }
     }
     StructType(schema.fields ++ outputFields)
   }
+
+  protected def extractInputDataType(schema: StructType, inputColName: String): Option[DataType] = {
+    val inputSplits = inputColName.split("\\.")
```

Comment:
I think you can use `UnresolvedAttribute.parseAttributeName`

Reply:
+1